kedro/io/kedro_data_catalog.py

```diff
@@ -398,7 +451,10 @@ def _get_dataset(
         return self.get_dataset(dataset_name, version, suggest)
 
     def add(
-        self, ds_name: str, dataset: AbstractDataset, replace: bool = False
+        self,
+        ds_name: str,
+        dataset: AbstractDataset | _LazyDataset,
```

Comment:
Are we sure we'd like to expose `_LazyDataset` to the users? If yes, then maybe we need to make it a proper `LazyDataset` that can be instantiated even outside of the catalog. Otherwise, I'd prefer for us to hide it from this method signature.

Reply:
We don't want to expose it and we added it here only until we use the old `add()` method for compatibility with the old catalog. I will go after the breaking change as setter will be used instead.